processor/src/parallel/core_trace_fragment/mod.rs

```diff
@@ -0,0 +1,1018 @@
+use core::ops::ControlFlow;
+
+use miden_air::{
+    Felt, FieldElement, RowIndex,
+    trace::{
+        STACK_TRACE_WIDTH, SYS_TRACE_WIDTH,
+        decoder::{NUM_OP_BITS, NUM_USER_OP_HELPERS},
+    },
+};
+use miden_core::{
+    ONE, OPCODE_PUSH, Operation, QuadFelt, StarkField, WORD_SIZE, Word, ZERO,
+    mast::{BasicBlockNode, MastForest, MastNode, MastNodeExt, MastNodeId, OpBatch},
+    precompile::PrecompileTranscriptState,
+    stack::MIN_STACK_DEPTH,
+    utils::range,
+};
+
+use crate::{
+    ContextId, ErrorContext, ExecutionError, ProcessState,
+    chiplets::CircuitEvaluation,
+    continuation_stack::Continuation,
+    decoder::block_stack::ExecutionContextInfo,
+    fast::{
+        NoopTracer, Tracer, eval_circuit_fast_,
+        trace_state::{
+            AdviceReplay, CoreTraceFragmentContext, ExecutionContextSystemInfo,
+            HasherResponseReplay, MemoryReadsReplay, NodeExecutionState,
+        },
+    },
+    host::default::NoopHost,
+    parallel::CORE_TRACE_WIDTH,
+    processor::{OperationHelperRegisters, Processor, StackInterface, SystemInterface},
+    utils::split_u32_into_u16,
+};
+
+mod execution;
+mod trace_row;
+
+// CORE TRACE FRAGMENT
+// ================================================================================================
+
+/// The columns of the main trace fragment. These consist of the system, decoder, and stack columns.
+///
+/// A fragment is a collection of columns of length `fragment_size` or less. Only a
+/// fragment containing a `HALT` operation is allowed to be shorter than
```

Comment:
Is it "only the last fragment which ends in a HALT operation is allowed to be shorter"?